src/b64.ml

```diff
@@ -16,77 +19,143 @@
-  let words = length / 4 in
-  let padding =
-    match length with
-    | 0 -> 0
-    | _ when input.[length - 2] = padding -> 2
-    | _ when input.[length - 1] = padding -> 1
-    | _ -> 0
-  in
-  let output = Bytes.make ((words * 3) - padding) '\000' in
-  for i = 0 to words - 1 do
-    let a = of_char ?alphabet input.[(4 * i) + 0]
-    and b = of_char ?alphabet input.[(4 * i) + 1]
-    and c = of_char ?alphabet input.[(4 * i) + 2]
-    and d = of_char ?alphabet input.[(4 * i) + 3] in
-    let n = (a lsl 18) lor (b lsl 12) lor (c lsl 6) lor d in
-    let x = (n lsr 16) land 255
-    and y = (n lsr 8) land 255
-    and z = n land 255 in
-    Bytes.set output ((3 * i) + 0) (char_of_int x) ;
-    if i <> words - 1 || padding < 2 then
-      Bytes.set output ((3 * i) + 1) (char_of_int y) ;
-    if i <> words - 1 || padding < 1 then
-      Bytes.set output ((3 * i) + 2) (char_of_int z)
-  done ;
-  Bytes.unsafe_to_string output
+exception Exists
+
+let padding_exists alphabet =
+  try String.iter (function '=' -> raise Exists | _ -> ()) alphabet; false
+  with Exists -> true
+
+let make_alphabet alphabet =
+  if String.length alphabet <> 64 then invalid_arg "Length of alphabet must be 64" ;
+  if padding_exists alphabet then invalid_arg "Alphabet can not contain padding character" ;
+  let emap = Array.init (String.length alphabet) (unsafe_get_uint8 alphabet)  in
+  let dmap = Array.make 255 none in
+  String.iteri (fun idx chr -> Array.unsafe_set dmap (Char.code chr) idx) alphabet ;
+  { emap; dmap; }
+
+let length_alphabet { emap; _ } = Array.length emap
+
+let default_alphabet = make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
+let uri_safe_alphabet = make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+
+let unsafe_set_be_uint16 =
+  if Sys.big_endian
+  then fun t off v -> unsafe_set_uint16 t off v
+  else fun t off v -> unsafe_set_uint16 t off (swap16 v)
+
+exception Out_of_bound
+
+let get_uint8 t off =
+  if off < 0 || off >= String.length t then raise Out_of_bound ;
+  unsafe_get_uint8 t off
+
+let padding = int_of_char '='
+
+let encode pad { emap; _ } input =
+  let n = String.length input in
+  let n' = n // 3 * 4 in
+  let res = Bytes.create n' in
+
+  let emap i = Array.unsafe_get emap i in
+
+  let emit b1 b2 b3 i =
+    unsafe_set_be_uint16 res i ((emap (b1 lsr 2 land 0x3f) lsl 8) lor (emap ((b1 lsl 4) lor (b2 lsr 4) land 0x3f))) ;
+    unsafe_set_be_uint16 res (i + 2) ((emap ((b2 lsl 2) lor (b3 lsr 6) land 0x3f) lsl 8) lor (emap (b3 land 0x3f))) in
+
+  let rec enc j i =
+    if i = n then ()
+    else if i = n - 1 then emit (unsafe_get_uint8 input i) 0 0 j
+    else if i = n - 2 then emit (unsafe_get_uint8 input i) (unsafe_get_uint8 input (i + 1)) 0 j
+    else
+    (emit
+       (unsafe_get_uint8 input i)
+       (unsafe_get_uint8 input (i + 1))
+       (unsafe_get_uint8 input (i + 2))
+       j ;
+     enc (j + 4) (i + 3)) in
+
+  let rec fix = function
```

Comment:
This also appears to be an unsafe interface as the argument isnt bounds checked. Should be `unsafe_fix`